File: 3108-minimum-cost-walk-in-weighted-graph/3108-minimum-cost-walk-in-weighted-graph.cpp
static auto init = []() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);
  return nullptr;
}();

constexpr unsigned int max_weight = (1 << bit_width(100'000u)) - 1u;

struct DisjointSet {
  DisjointSet* parent = this;
  unsigned int weight = max_weight;
};

DisjointSet& find(DisjointSet& x) {
  auto px = &x;
  while (px != px->parent) {
    tie(px, px->parent) = make_tuple(px->parent, px->parent->parent);
  }
  return *px;
}

void unify(DisjointSet& x, DisjointSet& y, unsigned int weight) {
  auto px = &find(x);
  auto py = &find(y);
  if (px == py) {
    if (px->weight) px->weight &= weight;
    return;
  }
  if (py->weight < px->weight) swap(px, py);
  if (px->weight) px->weight &= py->weight & weight;
  py->parent = px;
}

class Solution {
public:
  vector<int> minimumCost(int n, const vector<vector<int>>& edges, const vector<vector<int>>& query) {
    auto forest = vector<DisjointSet>(n);
    for (auto& edge : edges) {
      auto [u, v, w] = make_tuple(edge[0], edge[1], edge[2]);
      unify(find(forest[u]), find(forest[v]), w);
    }

    auto ans = vector<int>{};
    ans.reserve(size(query));
    for (auto& q : query) {
      auto [u, v] = make_pair(q[0], q[1]);
      const auto& node_u = find(forest[u]);
      const auto& node_v = find(forest[v]);
      ans.push_back(&node_u == &node_v ? node_u.weight : -1);
    }
    return ans;
  }
};

File: 3108-minimum-cost-walk-in-weighted-graph/3108-minimum-cost-walk-in-weighted-graph.py
class Solution:
    def minimumCost(self, n: int, edges: List[List[int]], query: List[List[int]]) -> List[int]:
        parents = [i for i in range(n)]
        rank = [1] * n
        minCost = [131071] * n

        def find(v):
            p = parents[v]
            while v != p:
                p, v = parents[p], p
            return p 

        def union(u, v, w):
            p1, p2 = find(u), find(v)                        
            minCost[p1] &= w & minCost[p2]
            minCost[p2] = minCost[p1]

            if p1 != p2: 
                if rank[p1] >= rank[p2]:
                    parents[p2] = p1 
                    rank[p1] += rank[p2]
                else:
                    parents[p1] = p2
                    rank[p2] += rank[p1]                                                
            else:
                return 

        for u, v, w in edges:
            union(u, v, w)     
        
        res = []
        for u, v in query:
            p1, p2 = find(u), find(v)
            if p1 == p2 and minCost[p1] != 131071:
                res.append(minCost[p1])
            else:
                res.append(-1)

        return res